src/res/mod.rs

```diff
@@ -244,124 +257,300 @@ impl Resources {
+    /// This also checks resources that were added after the last `maintain`.
+    /// If you only want to check maintained resources, use `has_maintained_value`.
+    /// If you only want to add a resource if it hasn't been added yet, use `entry`.
     pub fn has_value(&self, id: ResourceId) -> bool {
+        self.has_maintained_value(id) || self.new.lock().contains_key(&id)
+    }
+
+    /// Returns true if the specified resource type exists in `self`.
+    /// Note that this only checks if there's a maintained resource of that id.
+    pub fn has_maintained_value(&self, id: ResourceId) -> bool {
         self.resources.contains_key(&id)
     }
 
     /// Returns an entry for the resource with type `R` and id 0.
+    /// This calls `maintain` before creating the `Entry`.
     pub fn entry<R>(&mut self) -> Entry<R>
     where
         R: Resource,
     {
+        self.maintain();
+
         create_entry(self.resources.entry(ResourceId::new::<R>()))
     }
 
-    /// Fetches the resource with the specified type `T`.
-    /// The id is useful if you don't define your resources
-    /// in Rust or you want a more dynamic resource handling.
-    /// By default, the `#[derive(SystemData)]` passes `()`
-    /// as id.
+    /// Maintains the resources. This merges resources into the persistent
+    /// map. When `fetch`ing a resource that is not available, a `Default` will
+    /// be created. These values are inserted into a temporary map and get merged
+    /// on each `maintain`.
+    ///
+    /// The following methods are checking the temporary map:
+    ///
+    /// * `has_value`
+    /// * `try_fetch`
+    /// * `try_fetch_mut`
+    ///
+    /// The following methods call `maintain` at the beginning:
+    ///
+    /// * `add`
+    /// * `add_no_overwrite`
+    /// * `add_or_overwrite`
+    /// * `entry`
+    ///
+    /// These methods do not take care of unmaintained resources:
+    ///
+    /// * `has_maintained_value`
+    pub fn maintain(&mut self) {
+        self.resources
+            .extend(self.new.get_mut().drain().map(|(k, b)| (k, *b)));
+    }
+
+    /// Fetches the resource with the specified type `T` or calls `with` to
+    /// get a value which will be inserted.
+    ///
+    /// If the resource does not exist, the value returned by `with` will
+    /// be stored in a temporary map and fetched from there.
+    /// See `maintain` for more information.
     ///
     /// # Panics
     ///
     /// Panics if the resource is being accessed mutably.
-    /// Also panics if there is no such resource.
-    pub fn fetch<T>(&self) -> Fetch<T>
+    pub fn fetch<T, F, P>(&self, with: F) -> Fetch<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res(res_id, with()));
```

Comment:
You shouldn't call `with` here but in `def_res`

Reply:
Why? It'll definitely get called, so I don't see a reason to also add all the parameters to `def_res`.